syntax = "proto3";
package gdm;

service Gdm {
    rpc GetSamples(GetSamplesRequest) returns (GetSamplesResponse);
    rpc WriteSample(WriteSampleRequest) returns (WriteSampleResponse);
    rpc SignUp(SignUpRequest) returns (SignUpResponse);
    rpc SignIn(SignInRequest) returns (SignInResponse);
    rpc GetUser(GetUserRequest) returns (GetUserResponse);
    rpc GetDiagnosis(GetDiagnosisRequest) returns (GetDiagnosisResponse);
}

enum Ethnicity {
    GBR = 0;
    IND = 1;
    OTH = 2;  
}

message User{
    string user_id = 1;
    string password_hash = 2;
    int32 age = 3;
    float height = 4;
    float bmi = 5; 
    bool obesity = 6;
    Ethnicity ethnicity = 7;
    bool gravida = 8;
}

message Sample{
    int64 timestamp = 1;
    string user_id = 2;
    float ogtt1h = 3;
    float ogtt2h = 4;
}

message Samples{
    repeated Sample sample = 1;
}

message GetSamplesRequest {
    string user_id = 1;
}

message GetSamplesResponse{ 
    Samples samples = 1;
}

message WriteSampleRequest{
    Sample sample = 1;
}

message WriteSampleResponse{
    bool success = 1;
    string msg = 2;
}

message SignUpRequest {
    User user = 1;
}

message SignUpResponse {
    bool success = 1;
    string msg = 2;
}

message SignInRequest{
    string user_id = 1;
    string password_hash = 2;
} 

message SignInResponse{
    bool success = 1;
    string msg = 2;
}

message GetUserRequest {
    string user_id = 1;
}

message GetUserResponse {
    User user = 1;
}

message GetDiagnosisRequest{
    User user = 1;
    Sample sample = 2;
}

message GetDiagnosisResponse{
    bool hasGDM = 1;
}